### ISSUE TITLE
Function Sub-Typing: Parameters are Covariant but should be Contravariant

### ISSUE THREAD

**Issue by InnPatron** [InnPatron/smpl#1]
Let T, U be width constraint types.

T <: U iff forall fields f in U, f in T

Let F be a function of type A -> B
Let G be a function of type C -> D

Current behavior:
F <: G iff A <: C and B <: D

Fix:
F <: G iff C <: A and B <: D

Return values remain covariant. Parameters become contravariant.

Example:
```
fn foo(s: { x: int, y: int, z: int}) -> int {
  return x + y + z;
}

// Current behavior (Covariant)
let bar: fn ({x: int, y: int}) -> int = foo;
bar(init {x: 1, y: 1});
// Invalid b/c foo expects x, y, and z

// Proposed behavior (Contravariant)
let baz: fn({x: int, y: int, z: int, w: int}) -> int = foo;
baz(init {x: 1, y: 1, z: 1, w: 1});
// Fine b/c w field will be discarded
```

**Comment by InnPatron** [comment]
What happens when you pass a function expecting a nominal type in a parameter?

Width constraints are invariant to nominal record types. Should this restriction be altered for function parameters? Or should width constraints be allowed be used in the place of nominally typed records?

```
struct Point {
  x: int,
  y: int.
}

fn foo(p: Point) -> int { ... }

let f: ({x: int, y:int, z: int}) -> int = foo;  // Should this be allowed?

// Using anonymous struct where nominal type is expected is not allowed.
let pAnonymous: Point = init {
  x: 0, 
  y: 0,
};
```

**Comment by InnPatron** [comment]
Nominally-typed records may carry additional metadata at runtime. It may not be wise to allow the above scenario. However, allowing a smaller width constraint in the parameter where a nominal type was expected should be allowed. It should be trivial to convert the nominally-typed value to an anonymous record.

```
struct Point {
  x: int,
  y: int,
}

fn foo(p: {x :int}) -> int { ... }

let f: fn({x: int}) -> int = foo; 
let b: fn(Point) -> int = f;
b(init Point { x: 0, y: 0});
```

**Comment by InnPatron** [comment]
Final decision:

Function parameters are contravariant while return type remains covariant. This means:
- If a parameter is expected to be a nominal type, accept a narrower type. 
- If a parameter is expected to be a width constraint, accept a narrower type.
- If a parameter is expected to be a width constraint but a nominal type is provided, reject. Values of a nominal type may carry extra metadata which cannot be reconstructed from a general anonymous struct. Therefore, a parameter with type of an anonymous struct cannot be used where a nominal type is expected. This restriction holds elsewhere in SMPL.

The following code illustrates the point:
```
struct Point {
  x: int,
  y: int,
  z: int,
}

fn foo(p: {x: int}) -> int { ... }
fn bar(p: Point) -> int { ... }
fn qux(p: {x: int, y: int}) -> int { ... }

let allowed: fn(Point) -> int = foo; // Nominal parameter expected, narrower provided
let allowed: fn(Point) -> int = qux; // Nominal parameter expected, narrower provided
let allowed: fn({x: int, y: int}) -> int = foo; // Width constraint parameter expected, narrower provided
let allowed: fn({x: int, y: int}) -> int = qux; // Width constraint parameter expected, exact provided
let allowed: fn(Point) -> int = bar; // Nominal parameter expected, exact provided

let disallowed: fn({x: int}) -> int = qux; // Width constraint parameter expected, wider provided
let disallow: fn({x: int, y: int, z: int}) -> int = bar; // Width constraint parameter expected, nominal provided
```

**Comment by InnPatron** [comment]
Fixed by #4
